ValveResourceFormat/Resource/ResourceTypes/KV3.cs

```diff
@@ -0,0 +1,592 @@
+/*
+ * KV3Reader class.
+ * This class reads in valve KV3 files and stores them in its datastructure.
+ *  
+ * Interface:
+ *  KVFile file = KV3Reader.ParseKVFile( fileName );
+ *  String fileString = file.Serialize();
+ *  
+ * TODO:
+ *  - Test some more and find the bugs.
+ *  - Revisit state machine if bugs require it.
+ *  - Improve KVFile interface.
+ *  
+ * Author: Perry - https://github.com/Perryvw
+ * 
+ */
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.IO;
+using System.Globalization;
+
+namespace ValveKV
+{
+    public static class KV3Reader
+    {
+        private enum State {
+            HEADER,
+            SEEK_VALUE,
+            PROP_NAME,
+            VALUE_STRUCT,
+            VALUE_ARRAY,
+            VALUE_STRING,
+            VALUE_STRING_MULTI,
+            VALUE_NUMBER,
+            VALUE_BOOL,
+            VALUE_FLAGGED,
+            COMMENT,
+            COMMENT_BLOCK
+        }
+
+        private static char[] fileStream;
+        private static int index;
+
+        private static KVObject root = null;
+
+        private static String currentName;
+        private static StringBuilder currentString;
+
+        private static Stack<KVObject> objStack;
+        private static Stack<State> stateStack;
+
+        private const String whiteSpace = " \n\r\t";
+        private const String numerals = "-0123456789.";
+
+        public static KVFile ParseKVFile(string filename)
+        {
+            //Initialise datastructures
+            objStack = new Stack<KVObject>();
+            stateStack = new Stack<State>();
+            stateStack.Push(State.HEADER);
+
+            root = new KVObject("root");
+            objStack.Push(root);
+
+            currentString = new StringBuilder();
+
+            //Open stream
+            fileStream = File.ReadAllText(filename).ToCharArray();
```

Comment:
It was a stream first, that struggles with looking forward/back in C# though. Operating on a char array should realistically not make a difference.